### PR TITLE
Use mixins to modularize common behaviors across build-systems

### PR DESCRIPTION
**Pull request by alalazo**
This PR starts from a complaint from one of our users about `hdf5` compiler wrappers, that were still configured to use Spack wrappers. As I saw the code to filter those files repeated over and over in different packages I tried to factorize it in one place. Mixins seem a good enough approach to factor out common behavior that appears across different build systems.

### CODE REVIEW

File: lib/spack/spack/mixins.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    compiler_wrappers = []
+
+    #: phase after which the callback is invoked (default 'install')
+    filter_phase = 'install'
+
+    def __init__(self):
+
+        attr_name = '_InstallPhase_{0}'.format(self.filter_phase)
+
+        # Here we want to get the attribute directly from the class (noe from
```
Comment (davydden) [inline lib/spack/spack/mixins.py]: noe->not

File: lib/spack/spack/mixins.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class FilterCompilerWrappersPackageMixin(object):
+    """This mixin class registers a callback that filters a list of files
+    after installation and substitutes hardcoded paths pointing to the Spack
+    compiler wrappers with the corresponding 'real' compilers.
+    """
+
+    #: compiler wrappers to be filtered (needs to be overridden)
+    compiler_wrappers = []
```
Comment (davydden) [inline lib/spack/spack/mixins.py]: maybe rename to `filter_compiler_wrappers_in` so that it's more descriptive and easier to read when you see it being used.

Comment (davydden) [inline lib/spack/spack/mixins.py]: or `files_to_filter_compiler_wrappers`

Comment (alalazo) [inline lib/spack/spack/mixins.py]: Yeah, naming is kind of provisional... I am also not really satisfied by the name of the class. I think it would be fine if not exposed to packagers, but to somebody not used to programming `FilterCompilerWrappersPackageMixin` may be intimidating. I am definitely open for better names.

**Comment by alalazo** [comment]
For the record: I was able to compile a single source file application with `openmpi` and `mpich`. Also:
```console
$ /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14-x86_64/gcc-4.8/mpich-3.2-od7rbtncaf3ibcfatnhcbupngyc5nuar/bin/mpicxx -show
/usr/bin/g++-4.8 -I/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14-x86_64/gcc-4.8/mpich-3.2-od7rbtncaf3ibcfatnhcbupngyc5nuar/include -L/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14-x86_64/gcc-4.8/mpich-3.2-od7rbtncaf3ibcfatnhcbupngyc5nuar/lib -lmpicxx -Wl,-rpath -Wl,/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14-x86_64/gcc-4.8/mpich-3.2-od7rbtncaf3ibcfatnhcbupngyc5nuar/lib -lmpi
```
so it seems the wrappers have been filtered correctly.

**Comment by alalazo** [comment]
@tgamblin Do you have time to look at this one? If mixins are considered a viable solution I can try to see if there are other common patterns across packages. A likely candidate is try-compile a small executable that uses a library (e.g. `openblas`, `hdf5`)

**Comment by alalazo** [comment]
@tgamblin I'll try to see what I can do to improve the syntax using cloures or other features. At a first glance I see an issue with:
```python
@run_after('install')
filter_compiler_wrappers('mpicc', 'mpicxx')
```
because we miss the `def` keyword which I think it's needed by decorators.

**Comment by alalazo** [comment]
@tgamblin If you think it could be fine I can try to implement a declarative syntax much like the one we are using for directives. The result I have in mind is something like:
```python
class Hdf5(AutotoolsPackage):

    # Directives as usual
    version('1.10.1', '43a2f9466702fb1db31df98ae6677f15')
    version('1.10.0-patch1', '9180ff0ef8dc2ef3f61bd37a7404f295')
    version('1.10.0', 'bdc935337ee8282579cd6bc4270ad199')

    # ... then "mixins" with the same declarative syntax

    filter_compiler_wrappers(['h5c++', 'h5cc', 'h5fc'], after='install')

    # ...    
```
I can make `after = 'install` the default, so that it can be omitted in most cases. Let me know if this approach can be fine, as it may require a bit of work to insert hooks into the Package metaclass hierarchy.

**Comment by tgamblin** [comment]
@alalazo: This looks great. Can the compilers be varargs?  e.g.:

```
filter_compiler_wrappers('h5c++', 'h5cc', 'h5fc', after='install')
```

And could support for this be implemented as a special directive?  e.g., it seems like you could add some type of decorator to `directives.py` that would cause a function to be registered as a `run_after` hook.  So you could define it like this in `directives.py`:

```
@run_directive_after(default='install')
def filter_compiler_wrappers(*args):
    # do the actual filtering
```

The `run_directive_after` decorator would probably turn `filter_compiler_wrappers` into a class-style decorator, and it would add its own `kwarg` handling for the user's optional value... Would that work?

**Comment by tgamblin** [comment]
@alalazo: another question: should we call this `filter_compiler_wrappers`, or just `compiler_wrappers` or `compilers`?  The latter two are more declarative, so if there are other things we want to do to compiler wrappers later, we can just use the same annotation in the package.  Also do the compilers as this is currently written have to live in `bin`?  They could be other places like `libexec`...

**Comment by alalazo** [comment]
@tgamblin The mechanism to have a declarative syntax is in place now. I guess we need to sort out some details, such as where to put things (i.e. should `spack.mixins` stay?) or if names exposed to packagers are good enough.

>  Can the compilers be varargs?

Yes, and indeed they are now :smile: 

> And could support for this be implemented as a special directive? e.g., it seems like you could add some type of decorator to directives.py that would cause a function to be registered as a run_after hook. 

If you want to move the code that is now in `spack.mixins` elsewhere I am fine with it. I didn't put this stuff into `spack.directives` in the first place because the semantic is quite different. Directives are an implicit way to create an attribute in a package whereas 'mixins' will implicitly create a method and insert it in the appropriate position within the pipeline. I don't have strong opinions on where to put that code, so anything that you think is best for maintainability will do for me.

BTW, I'll continue to refer to these things as mixins, as they are conceptually mixins even though they don't use inheritance and the MRO.

> should we call this filter_compiler_wrappers, or just compiler_wrappers or compilers

I was about to use `compiler_wrappers`, but then stumbled on packages like `r` that needs to filter a file called `Makeconf`. So I opted for `filter_compiler_wrappers`. Just let me know which name you prefer, it won't be a problem to change it. Also @adamjstewart @davydden what do you think?

> Also do the compilers as this is currently written have to live in bin? They could be other places like libexec

No, they can be anywhere the `find` function can seek them! @adamjstewart I think I'll start using it more and more :+1:

**Comment by davydden** [comment]
> So I opted for filter_compiler_wrappers. Just let me know which name you prefer, it won't be a problem to change it. Also @adamjstewart @davydden what do you think?

I am in favour of more descriptive names like `filter_compiler_wrappers`. Luckily it's not a Fortran and we don't need to make 6 chars :smile:

**Comment by hartzell** [comment]
@alalazo -- speaking of weird corner cases...

I wonder if I'll be able to take advantage of this to clean up the [Perl package (around line 172)](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/perl/package.py#L172). 

It actually needs to fire up the installed `perl` and ask it where the files in question are located....

**Comment by alalazo** [comment]
> I wonder if I'll be able to take advantage of this

@hartzell If we can avoid calling `perl` maybe. If the method is `perl` specific I think there's not s much we can refactor...

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
ping

File: lib/spack/spack/mixins.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    def _filter_compiler_wrappers_impl(self):
+
+        tty.debug('Filtering compiler wrappers: {0}'.format(files))
+
+        # Compute the absolute path of the files to be filtered and
+        # remove links from the list.
+        abs_files = llnl.util.filesystem.find(self.prefix, files)
+        abs_files = [x for x in abs_files if not os.path.islink(x)]
+
+        kwargs = {'ignore_absent': True, 'backup': False, 'string': True}
```
Comment (tgamblin) [inline lib/spack/spack/mixins.py]: making the parameter explicit will make this not shadow it.

File: lib/spack/spack/mixins.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+        x = llnl.util.filesystem.FileFilter(*abs_files)
+
+        x.filter(os.environ['CC'], self.compiler.cc, **kwargs)
+        x.filter(os.environ['CXX'], self.compiler.cxx, **kwargs)
+        x.filter(os.environ['F77'], self.compiler.f77, **kwargs)
+        x.filter(os.environ['FC'], self.compiler.fc, **kwargs)
+
+        # Remove this linking flag if present (it turns RPATH into RUNPATH)
+        x.filter('-Wl,--enable-new-dtags', '', **kwargs)
```
Comment (tgamblin) [inline lib/spack/spack/mixins.py]: Some compilers now default to the new dtags, sadly.  I wish there were an easier way to handle this.  Might consider adding the *disable* flag here, but let's save that for another PR.

File: lib/spack/spack/mixins.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    PackageMixinsMeta.register_method_after(
+        _filter_compiler_wrappers_impl, after
```
Comment (tgamblin) [inline lib/spack/spack/mixins.py]: Could the code in `PackageMixinsMeta` and the mental requirements on mixin implementors be reduced by making this a bit more declarative?

Whoever writes a mixin has to know about `PackageMixinsMeta` as well as the methods on it, which do fairly boring stuff (add before, add after, etc).  What if they just wrote, say:

```python
@mixin
def filter_compiler_wrappers(*files, after='install'):
    def blah():
        # ... etc ...
    return {'after': {'install': {blah}}}
```

Then let the mixin decorator do its magic on the dict there.  Thoughts?  Seems easier to write docs for, in case we write docs for mixin function implementors.

File: lib/spack/spack/mixins.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    """Substitutes any path referring to a Spack compiler wrapper with the
+    path of the underlying compiler that has been used.
+
+    If this isn't done, the files will have CC, CXX, F77, and FC set to
+    Spack's generic cc, c++, f77, and f90. We want them to be bound to
+    whatever compiler they were built with.
+
+    Args:
+        *files: files to be filtered
+        **kwargs: at present supports the keyword 'after' to specify after
```
Comment (alalazo) [inline lib/spack/spack/mixins.py]: The fact that this parameter is a a `**kwargs` is mandated by the previous request of having the first parameter be a `*varargs`. To be more precise, what you ask is possible in `python@3`, but not in `python@2`. References [here](https://docs.python.org/2/reference/compound_stmts.html#function-definitions) and [here](https://docs.python.org/3.6/reference/compound_stmts.html#function-definitions) for your convenience.

---

Bottomline: here I can either turn the first parameter to be an iterable, and have an explicit keyword argument after it, or leave things like they are. Let me know which solution you prefer.

File: lib/spack/spack/mixins.py (hunk 3 above)
Comment (alalazo) [inline lib/spack/spack/mixins.py]: I'll change the name here.

File: lib/spack/spack/mixins.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            which phase the files should be filtered (defaults to 'install').
+    """
+    after = kwargs.get('after', 'install')
+
+    def _filter_compiler_wrappers_impl(self):
+
+        tty.debug('Filtering compiler wrappers: {0}'.format(files))
```
Comment (alalazo) [inline lib/spack/spack/mixins.py]: I'll remove the debug statement from here, and put `tty.debug` print in `llnl.util.filesystem.filter_file`.

File: lib/spack/spack/mixins.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+def filter_compiler_wrappers(*files, **kwargs):
+    """Substitutes any path referring to a Spack compiler wrapper with the
+    path of the underlying compiler that has been used.
+
+    If this isn't done, the files will have CC, CXX, F77, and FC set to
+    Spack's generic cc, c++, f77, and f90. We want them to be bound to
+    whatever compiler they were built with.
+
+    Args:
+        *files: files to be filtered
```
Comment (alalazo) [inline lib/spack/spack/mixins.py]: done

File: lib/spack/spack/mixins.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    after = kwargs.get('after', 'install')
+
+    def _filter_compiler_wrappers_impl(self):
+
+        tty.debug('Filtering compiler wrappers: {0}'.format(files))
+
+        # Compute the absolute path of the files to be filtered and
+        # remove links from the list.
+        abs_files = llnl.util.filesystem.find(self.prefix, files)
+        abs_files = [x for x in abs_files if not os.path.islink(x)]
```
Comment (alalazo) [inline lib/spack/spack/mixins.py]: Here I solved the issue in a slightly different manner. I maintained the call to `find`, but I added the possibility of specifying the root directory where to start the search. The rationale is that we have already seen weird packages that install stuff in subfolders that e.g. contain the architecture name or similar thing. I think this approach might be a good way to avoid both verbosity/hacks in those cases and performance issues or name clashes like you correctly pointed out.

File: lib/spack/spack/mixins.py (hunk 5 above)
Comment (alalazo) [inline lib/spack/spack/mixins.py]: I'll try to think of something here. If the API to packagers looks fine, can we merge this PR and start another to simplify the mixin implementation?

**Comment by alalazo** [comment]
Travis went green light, so for me this is ready for a further review.

**Comment by tgamblin** [comment]
@alalazo:

> Here I solved the issue in a slightly different manner. I maintained the call to find, but I added the possibility of specifying the root directory where to start the search. The rationale is that we have already seen weird packages that install stuff in subfolders that e.g. contain the architecture name or similar thing. I think this approach might be a good way to avoid both verbosity/hacks in those cases and performance issues or name clashes like you correctly pointed out.

I'm still sort of hesitant about this because:

1. I think it should be possible to specify relative paths for `filter_compiler_wrappers`, so that I can be precise about what to filter, and that's not currently possible.
2. There's still a relative root that defaults to the package root, so the default is likely slow.

RE:
> The rationale is that we have already seen weird packages that install stuff in subfolders that e.g. contain the architecture name or similar thing.

What if we just allow globbing (via `glob.glob`)?  e.g.:

```
    filter_compiler_wrappers('bin/*/some-script', 'lib/some-scriptname-*')
```

That makes it easy to handle most paths with weird platform/configuration-dependent things in them, and allows you to be precise.

File: lib/spack/spack/mixins.py (hunk 5 above)
Comment (tgamblin) [inline lib/spack/spack/mixins.py]: Yeah I think if the API for packagers looks ok we can go ahead.

**Comment by adamjstewart** [comment]
I agree with @tgamblin that we should support globs when it comes to finding files, but I also like the idea of specifying a directory where all of the files can be found (especially for OpenMPI).

**Comment by tgamblin** [comment]
@adamjstewart: does the interface I proposed handle the OpenMPI case?

> I think we should make people be specific and require prefix-relative paths.  For cases where that is a pain, we could add a keyword arg to set a different prefix, e.g.:
> 
> ```python
>     # vanilla
>     filter_compiler_wrappers('bin/mpicc', 'bin/mpic++', 'bin/mpifc')
> 
>     # with an arg:
>     filter_compiler_wrappers('mpicc', 'mpic++', 'mpifc', prefix='bin')
> ```

That doesn't do a `find`, though.  What about:

```python
    # vanilla
    filter_compiler_wrappers('bin/mpicc', 'bin/mpic++', 'bin/mpifc')

    # with an arg:
    filter_compiler_wrappers('mpicc', 'mpic++', 'mpifc', prefix='bin')

    # with globbing (for OpenMPI)
    filter_compiler_wrappers('share/openmpi/*-wrapper-data.txt')

    # with a prefix and globbing (for OpenMPI)
    filter_compiler_wrappers('*-wrapper-data.txt', prefix=''share/openmpi')

    # with find and a prefix (maybe not correct but you get the idea)
    filter_compiler_wrappers('mpi*', prefix=bin, spider=True)
```

The last one is a bit tricky; you would want to do os.path.walk on `prefix` and execute `glob.glob` there.

**Comment by adamjstewart** [comment]
@tgamblin That interface works for me. The last case you proposed there is actually the same as the `find` function. Maybe we should call it `recurse` to keep them in line?

**Comment by tgamblin** [comment]
@adamjstewart: `recurse` is fine with me, though I kind of like `spider` or `recursive` better b/c they're actually words 😛

**Comment by adamjstewart** [comment]
We can change `find/find_libraries/find_headers` to call it `recursive`.

**Comment by adamjstewart** [comment]
Probably need to update a few packages that call `find_libraries` to use `recursive` instead of `recurse`.

**Comment by alalazo** [comment]
In the last commit I changed find to accept glob expressions (which basically means I substituted a call to `fnmatch.fnmatch` with a corresponding one to `glob.glob`). Performance should not be an issue, as I have checked the implementation of `glob` and, apart a few conditionals to check in which case we fall, we pay extra only if we have directories in the list of files to be searched. This should cover all the cases above.

**Comment by alalazo** [comment]
@adamjstewart You're right. I'll wait for Travis to finish, because codecov seems to go crazy when there are two instances of Travis running on the same branch.

**Comment by adamjstewart** [comment]
> In the last commit I changed find to accept glob expressions (which basically means I substituted a call to `fnmatch.fnmatch` with a corresponding one to `glob.glob`). 

Actually, both `fnmatch` and `glob` accept glob expressions. The only difference appears to be how they treat hidden files (those starting with a dot).

**Comment by alalazo** [comment]
@adamjstewart Not only. `glob` understands things like `foo/*/*bar.txt` that `fnmatch` doesn't treat. In fact `glob` is required to be implemented using `fnmatch`.

**Comment by alalazo** [comment]
Is there something else that needs to be done here? 

The issues from previous reviews should be solved now. In particular:

>I think it should be possible to specify relative paths for filter_compiler_wrappers, so that I can be precise about what to filter, and that's not currently possible.

Now `find` accepts glob expressions, and `filter_compiler_wrappers` uses it in its implementation.

> There's still a relative root that defaults to the package root, so the default is likely slow.

The relative root is still the same, but now the default for the `filter_compiler_wrappers` is `recursive=False`

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by tgamblin** [comment]
@alalazo: I have one more minor tweak request.  Can we make the default relative paths and *not* search unless the user specifies `relative_root`?  I think it's unintuitive that this would search by default and we have to add some docs to `relative_root` to "strongly recommend setting this" because of it.  What if:

1. instead the default were to just specify files
2. rename `relative_root` to `search_from` or `search_root`, and search only if it is provided

This eliminates the need for the warning to prevent bad default behavior, as they only get search if they ask for it.  And they can make the relative root '/' if they want (to search from the prefix).

Sorry for being nitpicky -- does it seem better?

File: lib/spack/spack/mixins.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+            after
+                specifies after which phase the files should be
+                filtered (defaults to 'install')
+
+            relative_root
+                path relative to prefix where to start searching for
+                the files to be filtered. If not set the install prefix
+                wil be used as the search root. **It is highly recommended
+                to set this, as searching from the installation prefix may
+                affect performance severely in some cases**.
```
Comment (alalazo) [inline lib/spack/spack/mixins.py]: @tgamblin I started having a look at this, and have some questions.

> I think it's unintuitive that this would search by default and we have to add some docs to relative_root to "strongly recommend setting this" because of it.

If you refer to the lines above when you say  "strongly recommend setting this" they don't apply anymore, as we don't recurse by default. I think I need to remove the warning from the docstring (it was added when the default was to search recursively into subdirectories). 

> instead the default is just to specify relative paths, with no searching.

If we remove the use of `find` by default we'll also remove globbing - which was a previous request :smile: 

----

To summarize:

- right now we do a non-recursive globbing search by default, that starts at the `prefix` of the spec
- setting `relative_root` makes the search start from another location, relative to the prefix
- not setting `relative_root` could be a problem only for people that **explicitly** set a recursive search 

I would be for just renaming `relative_root` to `search_from` and update the docstring above, but I am not sure if I  am overlooking some of your concerns?

Comment (alalazo) [reply lib/spack/spack/mixins.py]: @tgamblin ping

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
ping
